libtsuba/src/AsyncOpGroup.cpp

```diff
@@ -0,0 +1,51 @@
+#include "tsuba/AsyncOpGroup.h"
+
+bool
+tsuba::AsyncOpGroup::FinishOne() {
+  auto op_it = pending_ops_.begin();
+  if (op_it == pending_ops_.end()) {
+    return false;
+  }
+  auto res = op_it->result.get();
+  if (!res) {
+    KATANA_LOG_DEBUG(
+        "async op for {} returned {}", op_it->location, res.error());
+    errors_++;
+    last_error_ = res.error();
+  } else {
+    res = op_it->on_complete();
+    if (!res) {
+      KATANA_LOG_DEBUG(
+          "complete cb for async op for {} returned {}", op_it->location,
+          res.error());
+    }
+  }
+  pending_ops_.erase(op_it);
+  return true;
+}
+
+katana::Result<void>
+tsuba::AsyncOpGroup::Finish() {
+  while (FinishOne()) {
+    // Wait for all ops
+  }
+
+  if (errors_ > 0) {
+    return last_error_.error().WithContext(
+        "{} of {} async write ops returned errors", errors_, total_);
+  }
+
+  return last_error_;
+}
+
+void
+tsuba::AsyncOpGroup::AddOp(
+    std::future<katana::Result<void>> future, std::string file,
+    const std::function<katana::Result<void>()>& on_complete) {
+  pending_ops_.emplace_back(AsyncOp{
+      .result = std::move(future),
+      .location = std::move(file),
```

Comment:
Seeing dot-based structure initialization combined with std::move is very steampunk.

Reply:
⚙️ 😎